flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,211 @@
+                        (KnnModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+            }
+            DenseVector feature = (DenseVector) row.getField(featureCol);
+            double prediction = predict(feature);
+            return Row.join(row, Row.of(prediction));
+        }
+
+        private double predict(DenseVector inputFeature) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue =
+                    new PriorityQueue<>(Comparator.comparingDouble(o -> -o.f0));
+            double featureNorm = 0.0;
+            for (int i = 0; i < inputFeature.size(); ++i) {
+                featureNorm += inputFeature.values[i] * inputFeature.values[i];
+            }
+            Tuple2<DenseVector, Double> sample = Tuple2.of(inputFeature, featureNorm);
+            double[] labelValues = knnModelData.labels.values;
+            double[] normValues = knnModelData.featureNorms.values;
+            DenseVector distanceVector = new DenseVector(labelValues.length);
+            BLAS.gemv(-2.0, knnModelData.packedFeatures, true, sample.f0, 0.0, distanceVector);
+            for (int i = 0; i < distanceVector.values.length; i++) {
+                distanceVector.values[i] =
+                        Math.sqrt(Math.abs(distanceVector.values[i] + sample.f1 + normValues[i]));
+            }
+            List<Tuple2<Double, Double>> distanceAndLabels = new ArrayList<>(labelValues.length);
+            for (int i = 0; i < labelValues.length; i++) {
+                distanceAndLabels.add(Tuple2.of(distanceVector.values[i], labelValues[i]));
+            }
+            for (Tuple2<Double, Double> distanceAndLabel : distanceAndLabels) {
+                if (priorityQueue.size() < k) {
+                    priorityQueue.add(Tuple2.of(distanceAndLabel.f0, distanceAndLabel.f1));
+                } else {
+                    Tuple2<Double, Double> head = priorityQueue.peek();
+                    if (priorityQueue.comparator().compare(head, distanceAndLabel) < 0) {
+                        Tuple2<Double, Double> peek = priorityQueue.poll();
+                        if (peek != null) {
+                            peek.f0 = distanceAndLabel.f0;
+                            peek.f1 = distanceAndLabel.f1;
+                            priorityQueue.add(peek);
+                        }
+                    }
+                }
+            }
+            List<Double> labels = new ArrayList<>();
+            while (!priorityQueue.isEmpty()) {
+                Tuple2<Double, Double> distanceAndLabel = priorityQueue.poll();
+                labels.add(distanceAndLabel.f1);
+            }
+            double percent = 1.0 / labels.size();
+            Map<Double, Double> labelWeights = new HashMap<>(0);
+            for (Double label : labels) {
+                labelWeights.merge(label, percent, Double::sum);
+            }
+            double maxWeight = 0.0;
+            double prediction = 0.0;
```

Comment:
It seems that `label` or `predictedLabel` would be slightly more self-explanatory. And maybe rename this method from `predict` to `predictLabel` to be more consistent with `PredictLabelFunction`. What do you think?

Reply:
OK